src/IK/src/QPInverseKinematics.cpp

```diff
@@ -0,0 +1,405 @@
+            m_pimpl->equalityConstraints.push_back(m_pimpl->tasks[taskName]);
+    } else
+    {
+        m_pimpl->costs.push_back(m_pimpl->tasks[taskName]);
+    }
+
+    return true;
+}
+
+bool QPInverseKinematics::finalize(const System::VariablesHandler& handler)
+{
+    constexpr auto logPrefix = "[QPInverseKinematics::finalize]";
+
+    //
+    std::size_t numberOfConstraints = 0;
+    for (auto& [name, task] : m_pimpl->tasks)
+    {
+        if (!task.task->setVariableHandler(handler))
+        {
+            log()->error("{} Error while setting the variable handler in the task named {}, having "
+                         "the following description {}.",
+                         logPrefix,
+                         name,
+                         task.task->getDescription());
+            return false;
+        }
+
+        if (task.priority == 0)
+        {
+            numberOfConstraints += task.task->size();
+        }
+    }
+
+    m_pimpl->solver.data()->setNumberOfVariables(handler.getNumberOfVariables());
+    m_pimpl->solver.data()->setNumberOfConstraints(numberOfConstraints);
+
+    // resize matrices
+    m_pimpl->hessian.resize(handler.getNumberOfVariables(), handler.getNumberOfVariables());
+    m_pimpl->gradient.resize(handler.getNumberOfVariables());
+
+    m_pimpl->constraintMatrix.resize(numberOfConstraints, handler.getNumberOfVariables());
+    m_pimpl->upperBound.resize(numberOfConstraints);
+    m_pimpl->lowerBound.resize(numberOfConstraints);
+
+    if (!handler.getVariable(m_pimpl->robotVelocityVariable.name,
+                             m_pimpl->robotVelocityVariable))
+    {
+        log()->error("{} Error while retrieving the robot velocity variable.", logPrefix);
+        return false;
+    }
+
+    return true;
+}
+
+std::vector<std::string> QPInverseKinematics::getTasksName() const
+{
+    std::vector<std::string> tasksName;
+
+    for (const auto& [name, task] : m_pimpl->tasks)
+    {
+        tasksName.push_back(name);
+    }
+
+    return tasksName;
+}
+
+bool QPInverseKinematics::advance()
+{
+    constexpr auto logPrefix = "[QPInverseKinematics::advance]";
+
+    // when advance is called the previous solution is no more valid
+    m_pimpl->isValid = false;
+
+    if (!m_pimpl->isInitialized)
+    {
+        log()->error("{} Please call initialize() before advance().", logPrefix);
+        return false;
+    }
+
+    // update of all the tasks
+    for (auto& [name, task] : m_pimpl->tasks)
+    {
+        task.task->update();
+    }
+
+    // Compute the gradient and the hessian
+    m_pimpl->hessian.setZero();
+    m_pimpl->gradient.setZero();
+    for (const auto& cost : m_pimpl->costs)
+    {
+        Eigen::Ref<const Eigen::MatrixXd> A = cost.get().task->getA();
+        Eigen::Ref<const Eigen::VectorXd> b = cost.get().task->getB();
+
+        m_pimpl->hessian += A.transpose() * cost.get().weight.asDiagonal() * A;
+        m_pimpl->gradient -= A.transpose() * cost.get().weight.asDiagonal() * b;
```

Comment:
I think we can avoid to add a buffer following what is written here http://eigen.tuxfamily.org/dox/TopicWritingEfficientProductExpression.html
```suggestion
        m_pimpl->hessian.noalias() += A.transpose() * cost.get().weight.asDiagonal() * A;
        m_pimpl->gradient.noalias() -= A.transpose() * cost.get().weight.asDiagonal() * b;
```

Reply:
I think that works only if ``weight`` is a scalar, i.e.
```
m_pimpl->hessian.noalias() += cost.get().weight * (A.transpose() * A);
```
so that we fall in the second row of the table in http://eigen.tuxfamily.org/dox/TopicWritingEfficientProductExpression.html. Otherwise, it is a three matrix multiplication, and I think that still requires a temporary object.

Reply:
I investigated a bit in the solution you propose. I took the approach from https://stackoverflow.com/questions/52918480/identifying-temporary-object-creation-in-eigen

The following code prints a string every time a dynamic allocation is done. 

```cpp
#include <cassert>
#include <iostream>

static long int nb_temporaries;
inline void on_temporary_creation(long int size) {
  // here's a great place to set a breakpoint when debugging failures in this test!
  if(size!=0) nb_temporaries++;

  std::cerr << "Number of allocation " << nb_temporaries << std::endl ;
}

#define EIGEN_DENSE_STORAGE_CTOR_PLUGIN { on_temporary_creation(size); }

#include <Eigen/Dense>

int main()
{
    Eigen::MatrixXd a(10,10);
    Eigen::MatrixXd c(10,10);
    Eigen::VectorXd weight(10);

    c.setIdentity();
    weight.setRandom();
    a.setRandom();


    // option 1
    Eigen::MatrixXd temp(10,10);
    temp.noalias() = a.transpose() * weight.asDiagonal() * a;
    c.noalias() += temp;

    // option 2
    // c.noalias() += a.transpose() * weight.asDiagonal() * a;

    return 0;
}
```

If I run the code with `option 1` enabled I got
```
Number of allocation 1
Number of allocation 2
Number of allocation 3
Number of allocation 4
Number of allocation 5
```

Allocation 1 is for `a`, allocation 2 is for `c`, allocation 3 if for `weight` and allocation 4 is for `temp`. (this dynamic allocation are done only once in our case). Allocation 5 is a temporary object created in `temp.noalias() = a.transpose() * weight.asDiagonal() * a;` Thus every cicle eigen allocate a matrix.

If I run the code with `option 2` enabled I got
```
Number of allocation 1
Number of allocation 2
Number of allocation 3
Number of allocation 4
```
Allocation 1 is for `a`, allocation 2 is for `c`, allocation 3 if for `weight`. allocation 4 is a temporary object created in
`c.noalias() += a.transpose() * weight.asDiagonal() * a;`

Reply:
This is a possible solution 
```
    Eigen::MatrixXd temp(10,10);                                                                     
    temp.noalias() = a.transpose() * weight.asDiagonal();
    c.noalias() += temp * a;
```

In this case we have 
```
Number of allocation 1
Number of allocation 2
Number of allocation 3
Number of allocation 4
```

Reply:
This patch goes in the direction you propose. However it breaks the symmetry between cost and constraints (`std::vector<std::reference_wrapper<const TaskWithPriority>> costs` -> `std::vector<std::reference_wrapper<TaskWithPriority>> costs`) (i.e. the `const` disappear)


```patch
diff --git a/src/IK/src/QPInverseKinematics.cpp b/src/IK/src/QPInverseKinematics.cpp
index f85b8d8f..6845d00d 100644
--- a/src/IK/src/QPInverseKinematics.cpp
+++ b/src/IK/src/QPInverseKinematics.cpp
@@ -21,6 +21,8 @@ struct QPInverseKinematics::Impl
         std::shared_ptr<Task> task;
         std::size_t priority;
         Eigen::VectorXd weight;
+
+        Eigen::MatrixXd tmp;
     };
 
     IKState solution;
@@ -32,7 +34,7 @@ struct QPInverseKinematics::Impl
     std::unordered_map<std::string, TaskWithPriority> tasks;
 
     std::vector<std::reference_wrapper<const TaskWithPriority>> constraints;
-    std::vector<std::reference_wrapper<const TaskWithPriority>> costs;
+    std::vector<std::reference_wrapper<TaskWithPriority>> costs;
 
     Eigen::MatrixXd hessian;
     Eigen::VectorXd gradient;
@@ -207,6 +209,12 @@ bool QPInverseKinematics::finalize(const System::VariablesHandler& handler)
         }
     }
 
+    // resize the  temporary matrix usefull to reduce dynamics allocation while advance is called
+    for (auto& cost : m_pimpl->costs)
+    {
+        cost.get().tmp.resize(handler.getNumberOfVariables(), cost.get().weight.size());
+    }
+
     m_pimpl->solver.data()->setNumberOfVariables(handler.getNumberOfVariables());
     m_pimpl->solver.data()->setNumberOfConstraints(numberOfConstraints);
 
@@ -262,13 +270,15 @@ bool QPInverseKinematics::advance()
     // Compute the gradient and the hessian
     m_pimpl->hessian.setZero();
     m_pimpl->gradient.setZero();
-    for (const auto& cost : m_pimpl->costs)
+    for (auto& cost : m_pimpl->costs)
     {
         Eigen::Ref<const Eigen::MatrixXd> A = cost.get().task->getA();
         Eigen::Ref<const Eigen::VectorXd> b = cost.get().task->getB();
 
-        m_pimpl->hessian += A.transpose() * cost.get().weight.asDiagonal() * A;
-        m_pimpl->gradient -= A.transpose() * cost.get().weight.asDiagonal() * b;
+        cost.get().tmp.noalias() = A.transpose() * cost.get().weight.asDiagonal();
+
+        m_pimpl->hessian.noalias() += cost.get().tmp * A;
+        m_pimpl->gradient.noalias() -= cost.get().tmp * b;
     }
 
     // compute the constraints
```

Reply:
Done here 852c792b33903e8fa49f024420fb8e50c28847da